File: 2023-11-23-revisao-de-python/09-conjuntos-de-produtos-que-resultam-em-n.py
# Projete uma função que retorne, para um dado número inteiro
# n, uma lista contento todos os pares de números inteiros cujo o
# produto resulta em n.

# Por simplicidade, considere que o par (x, y) é diferente de (y, x)
def conjuntos(n) -> list[set[int]]:
    '''
        Dado número inteiro *n*, a função retorna uma lista contento todos 
        os conjuntos(conjuntos) de números inteiros cujo o produto resulta em *n*.
        Exemplos
        >>> pares(5)
        [{1, 5}]
        >>> pares(10)
        [{1, 10}, {2, 5}]
        >>> pares(15)
        [{1, 15}, {3, 5}]
        >>> pares(150)
        [{1, 150}, {2, 75}, {50, 3}, {5, 30}, {25, 6}, {10, 15}]
    '''
    lista: list[set[int]] = []

    for i in range(1, n+1):
        for j in range(1, n+1):
            produtoEhIgualAhN = (i * j) == n

            if produtoEhIgualAhN:
                conjunto = set()
                conjunto.add(i)
                conjunto.add(j)

                if conjunto not in lista:
                    lista.append(conjunto)

    return lista

def main():
    n: int = int(input("Digite um número inteiro: "))

    print(f"Os conjuntos de inteiros que resultam em {n} são: {conjuntos(n)}")

if __name__ == "__main__":
    main()

File: 2023-11-23-revisao-de-python/09-conjuntos-de-produtos-que-resultam-em-n.c
// # Projete uma função que retorne, para um dado número inteiro
// # n, uma lista contento todos os pares de números inteiros cujo o
// # produto resulta em n.

// # Por simplicidade, considere que o par (x, y) é diferente de (y, x)
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int primeiro;
    int segundo;
} Conjunto;

Conjunto* conjuntos(int n) {
    Conjunto* lista = malloc(n * sizeof(Conjunto));
    int size = 0;

    for ( int i = 1; i <= n; i++ ) {
        for ( int j = 1; j <= n; j++ ) {
            if ( i * j != n ) continue;

            lista[size].primeiro = i;
            lista[size].segundo = j;
            size++;
        }
    }

    return lista;
}

int main() {
    Conjunto* conjuntosDeN;
    conjuntosDeN = conjuntos(5); // [{primeiro: 1, segundo: 5}]
    printf("[{%d, %d}]\n", conjuntosDeN[0].primeiro, conjuntosDeN[0].segundo);

    printf("\n");
    conjuntosDeN = conjuntos(150); // [{1, 150}, {2, 75}, {50, 3}, {5, 30}, {25, 6}, {10, 15}]
    for ( int i = 0; i < 6; i++ ) {
        if ( i == 0 ) printf("[ ");
        printf("{%d, %d} ", conjuntosDeN[i].primeiro, conjuntosDeN[i].segundo);
        if ( i == 5 ) printf("]\n");
    }
    //     >>> pares(10)
    //     [{1, 10}, {2, 5}]
    //     >>> pares(15)
    //     [{1, 15}, {3, 5}]
    //     >>> pares(150)
    //     

    // n: int = int(input("Digite um número inteiro: "))
    // print(f"Os conjuntos de inteiros que resultam em {n} são: {conjuntos(n)}")
}
